test/jdk/sun/security/ssl/SSLSocketImpl/CloseSocket.java

```diff
@@ -25,26 +25,59 @@
  * @test
  * @bug 4674913
  * @summary Verify that EOFException are correctly handled during the handshake
+ * @library /javax/net/ssl/templates
  * @author Andreas Sterbenz
  * @run main/othervm CloseSocket
  */
 
-import javax.net.SocketFactory;
 import javax.net.ssl.SSLSocket;
-import javax.net.ssl.SSLSocketFactory;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
-import java.net.ServerSocket;
-import java.net.Socket;
 import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
 
-public class CloseSocket {
 
-    private static ArrayList<TestCase> testCases = new ArrayList<>();
+public class CloseSocket extends SSLSocketTemplate {
 
-    static {
-        testCases.add(socket -> socket.startHandshake());
+    private static Thread clientThread = null;
```

Comment:
Shouldn't this variable should be `volatile`? If I'm not mistaken it's set in one thread and potentially read in a different thread? An alternative could be to use a CountDownLatch instead.

Reply:
Thanks, updated to be volatile.

SSLSocket::startHandshake internally first checks that the socket is not closed or broken and still connected, so it needs the server to close the socket after those verifications are performed to reproduce the test scenario, thus a CountDownLatch in the test before calling startHandshake would not guarantee that its internal operations are run before the server, already unblocked at that time, closes the socket

A CountDownLatch after startHandshake does not work either since the client keeps waiting for a server response, which is blocked waiting for the latch. That is why I think that looking at the thread stack is the best way to guarantee the scenario is properly verified

Reply:
OK

Reply:
Can you please add this as a comment in a test file for clientThread variable? I am sure we will be trying to convert it to CountDownLatch in future.